lib/ecto/adapters/postgres/connection.ex

```diff
@@ -876,6 +885,10 @@ if Code.ensure_loaded?(Postgrex) do
       ["NOT (", expr(expr, sources, query), ?)]
     end
 
+    defp expr({bnot, _, [expr]}, sources, query) when bnot in ~w(~~~ bnot)a do
+      ["~" | maybe_paren(expr, sources, query)]
```

Comment:
```suggestion
     ["~ (", expr(expr, sources, query), ?)]
```

I think we can just always parenthesize it, like the logical not. And then we can add a clause for these in `maybe_paren`:

```elixir
defp maybe_paren({bnot, _, [_]} = expr, sources, query) when bnot in ~w(~~~ bnot)a,
      do: paren_expr(expr, sources, query)
```

Then reason we want this clause is for casting. For instance, without this clause we could get this:

```sql
~B'10001'::integer
```

Which postgres will change to 

```sql
~17
```

And then it will choose whatever bit representation it thinks is best and we lose the one we gave it. We want it to do this instead

```sql
(~B'10001')::integer
```

Reply:
Fixed this.

I changed this:
```elixir
defp expr({bnot, _, [expr]}, sources, query)
```

to always parenthesize it in both MySQL and Postgres connectors, as you suggested.

I also added a case for `bnot` to `maybe_paren` in the PostgreSQL connector and to `op_to_binary` in the MySQL connector, although the latter doesn't seems to be necessary.

Though, I sliglthy prefer to not always parenthesize the `bnot` expression, I find the generated SQL more readable. For instance, with the current version your example becomes

```elixir
(~(B'10001'))::integer
```

While with the previous version it would have been:


```elixir
(~B'10001')::integer
```

Reply:
We can do it that way if we are sure it won't lead to precedence issues. I am not confident enough to say either way atm. I would have to think about the possible situations and why `not` is always parenthesized. It's possible the two are not similar enough. Or maybe `maybe_paren` would work for `not` as well.